linkerd/opencensus/src/lib.rs

```diff
@@ -56,87 +61,111 @@ where
         }
     }
 
-    async fn run(mut self) {
-        'reconnect: loop {
-            let (request_tx, request_rx) = mpsc::channel(1);
-            let mut svc = TraceServiceClient::new(self.client.clone());
-            let req = grpc::Request::new(request_rx);
+    async fn run(self) {
+        let Self {
+            client,
+            node,
+            mut spans,
+            mut metrics,
+        } = self;
+
+        // Holds the batch of pending spans. Cleared as the spans are flushed.
+        // Contains no more than MAX_BATCH_SIZE spans.
+        let mut accum = Vec::new();
+
+        let mut svc = TraceServiceClient::new(client);
+        loop {
             trace!("Establishing new TraceService::export request");
-            self.metrics.start_stream();
-            let mut rsp = Box::pin(svc.export(req));
-            let mut drive_rsp = true;
-
-            loop {
-                tokio::select! {
-                    res = &mut rsp, if drive_rsp => match res {
-                        Ok(_) => {
-                            drive_rsp = false;
-                        }
-                        Err(error) => {
-                            tracing::debug!(%error, "response future failed, sending a new request");
-                            continue 'reconnect;
-                        }
-                    },
-                    res = self.send_batch(&request_tx) => match res {
-                        Ok(false) => return, // The span strean has ended --- the proxy is shutting down.
-                        Ok(true) => {} // Continue the inner loop and send a new batch of spans.
-                        Err(()) => continue 'reconnect,
+            let (tx, rx) = mpsc::channel(1);
+            match svc.export(grpc::Request::new(rx)).await {
```

Comment:
I don't think this is right --- the future returned by `export` finishes when the RPC is *completed*. a lot of the complexity from the previous impl came from the need to drive this future concurrently with exporting spans --- i think if we just `await` this, the stream will have closed by the time we actually start writing spans to the stream, and no spans will ever be exported.

Reply:
@hawkw updated; PTAL